### ISSUE TITLE
[Question] How can I prevent opening app when press on send icon on REPLY?

### ISSUE THREAD

**Issue by husen-hn** [rafaelsetragni/awesome_notifications#265]
How can I prevent opening the app when I receive notifications from background/terminated from FCM?
I just want to reply to messages via notifications. No more (I do not want to open the whole app)

```Dart
AwesomeNotifications().actionStream.listen((receivedNotification) async {
      if (!StringUtils.isNullOrEmpty(receivedNotification.buttonKeyInput)) {
        processInputTextReceived(receivedNotification);
      }
}
```

**Comment by rafaelsetragni** [comment]
Im developing right now a new version of Awesome Notifications to allow to receive actions without open the app, or even ignore the action.

This feature will be ready to use in version 0.0.7. You can follow our progress with the [pull request #257](https://github.com/rafaelsetragni/awesome_notifications/pull/257)

**Comment by husen-hn** [comment]
> Im developing right now a new version of Awesome Notifications to allow to receive actions without open the app, or even ignore the action.
> 
> This feature will be ready to use in version 0.0.7. You can follow our progress with the [pull request #257](https://github.com/rafaelsetragni/awesome_notifications/pull/257)

Thank you very much for your efforts
I wrote this code 
‍‍‍```dart
AwesomeNotifications()
      .actionStream
      .listen((ReceivedAction receivedNotification) async{})
```
before `runApp(MyApp())`, but again the problem has not been solved yet. 

Also, when will version 0.0.7 be released? Yesterday I saw that version 0.0.7 was released but today I saw that it has changed to version 0.0.6 + 10

**Comment by gbbest15** [comment]
i also cant use this plugin untill the version 0.0.7 is release.. i watched the youtube video of the changes for the 0.0.7.. i am waiting for the release @husen-hn

**Comment by husen-hn** [comment]
related to this [answer](https://github.com/rafaelsetragni/awesome_notifications/pull/257#issuecomment-914801436) 0.0.7 is ready for Android.
can anyone explain me...
How can I use version 0.0.7 right now? (I cloned this repo and added it to project, I don't know, this way is right or not)
How can I use SilentBackgroundAction in code?
couse of you don't use these solutions in [example app](https://github.com/rafaelsetragni/awesome_notifications/tree/master/example), I need a solution with code examples.

**Comment by rafaelsetragni** [comment]
The 0.0.7 version will only be available officially when both platforms are considered finished.
But you can use it right now importing the code from the development branch.

Add this into your `pusbpec.yaml`:
```yaml
# awesome_notifications: ^0.0.6+10
awesome_notifications:
     git:
       url: https://github.com/rafaelsetragni/awesome_notifications.git
       ref: 441fabd
```

and add after `description` the tag bellow to enable git references into `pusbec.yaml`:

```yaml
description: lalalalalalala
publish_to: 'none' # Remove this line if you wish to publish to pub.dev
```

This git link is a direct reference to the branch [notifications-isolates](https://github.com/rafaelsetragni/awesome_notifications/tree/notifications-isolates)

And there's no documentation on version 0.0.7 either because it's not finished yet, but i will provide above a fast step tutorial to guide you on it. You can also check the example folder to see how everything is working on the new version.

1 - Initialize the plugin with `AwesomeNotifications.initialize()`
2 - Declare in the App class (the first class after main method) the four static methods to handle notifications and the global navigator key:

```Dart
class App extends StatefulWidget {
  App();

  static final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();

  static String name = 'Awesome Notifications - Example App';
  static Color mainColor = Color(0xFF9D50DD);

  /// Use this method to detect when a new notification or a schedule is created
  static Future <void> onCreatedNotificationMethod(ReceivedNotification receivedNotification) async {
    /// This replaces createdStream
  }

  /// Use this method to detect every time that a new notification is displayed
  static Future <void> onDisplayedNotificationMethod(ReceivedNotification receivedNotification) async {
    /// This replaces displayedStream
  }

  /// Use this method to detect if the user dismissed a notification
  static Future <void> onDismissedNotificationMethod(ReceivedAction receivedAction) async {
    /// This replaces dismissedStream
  }

  /// Use this method to detect when the user taps on a notification or action button
  static Future <void> onActionNotificationMethod(ReceivedAction receivedAction) async {{
    /// This replaces actionStream

    switch(receivedAction.notificationActionType){

      // Default Notification behavior, that brings the app to foreground
      case NotificationActionType.BringToForeground:
        // your code goes here
        break;

      // Do not bring the app to foreground and do not work with app killed
      case NotificationActionType.KeepOnTopAction:
        // your code goes here
        break;

      // Receive silent notification actions without bring the app to foreground
      case NotificationActionType.SilentAction: // Runs on main thread
      case NotificationActionType.SilentBackgroundAction: // Runs on background thread (without UI)
        // your code goes here
        return;

      // disabled actions never reach this point, they die
      // at native level.
      case NotificationActionType.DisabledAction:
      default:
        break;
    }

    // Navigate into pages, avoiding to open the notification details page over another details page already opened
    App.navigatorKey.currentState?.pushNamedAndRemoveUntil(targetPage,
            (route) => (route.settings.name != targetPage) || route.isFirst,
        arguments: receivedAction);
  }

  @override
  _AppState createState() => _AppState();
}
```

3 - Pass the global navigation key into your MaterialApp class (Normaly inside _AppState class):

```Dart

class _AppState extends State<App> {

  @override
  void dispose() {
    AwesomeNotifications().dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      // The navigator key is necessary to allow to navigate through static classes
      navigatorKey: App.navigatorKey,

      title: App.name,
      color: App.mainColor,

      // These routes are declared in a separated file called routes.dart
      initialRoute: PAGE_HOME,
      routes: materialRoutes
    );
  }
}
```
 4 - And finally after Material Page, initialize your static methods to start to capture the notifications.

```Dart

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() =>
      _HomePageState();
}

class _HomePageState extends State<HomePage> {

  @override
  void initState() {
    super.initState();

    // Necessary to request the user permission to send notifications
    requestNotificationPermissions();

    // Only after set at least one method, the notification's events are delivered
    AwesomeNotifications().setListeners(
        onCreatedNotificationMethod:   App.onCreatedNotificationMethod,
        onDisplayedNotificationMethod: App.onDisplayedNotificationMethod,
        onActionNotificationMethod:    App.onActionNotificationMethod,
        onDismissedNotificationMethod: App.onDismissedNotificationMethod
    );
  }

  Future<bool> requestNotificationPermissions() async {

    bool isAllowed = await requireUserNotificationPermissions(context);

    if(mounted){
      setState(() {
        notificationsAllowed = isAllowed;
      });
    } else {
      notificationsAllowed = isAllowed;
    }
    return isAllowed;
  }

  @override
  void dispose() {
    WidgetsBinding.instance!.removeObserver(this);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: SizedBox()
    );
  }
}
```

Test it yourself and keep in touch if you encounter any problems or if everything is working just fine.

**Comment by husen-hn** [comment]
Thank. The problem is really solved.
I don't have any problem when I set `NotificationActionType.BringToForeground`.
and when I set `NotificationActionType.SilentAction` or `NotificationActionType.SilentBackgroundAction`, I don't receive any data inside `onActionNotificationMethod()` when app is running on foreground or background anyway I don't receive any data.
but when I set `NotificationActionType.KeepOnTopAction` its worked on foreground and background but not in terminated.
```dart
await AwesomeNotifications().createNotification(
      content: NotificationContent(
        id: id,
        channelKey: 'private_channel',
        ticker: username,
        title: username,
        body: message,
        color: AppColors.primaryColor,
      ),
      actionButtons: [
        NotificationActionButton(
          key: 'REPLY',
          label: 'reply',
          requireInputText: true,
          notificationActionType: NotificationActionType.KeepOnTopAction,
        ),
]);
```

```dart
static Future<void> onActionNotificationMethod(
      ReceivedAction receivedAction) async {
    debugPrint('$receivedAction')
}
```

**Comment by rafaelsetragni** [comment]
Sorry @husen-hn ! My bad! I forgot to update properly the object `Received Action`.
I swear that i did it (i remember that) but it's not there. I guess i gonna need some rest after 0.0.7 version....

Update your `pusbpec.yaml` as bellow:
```yaml
# awesome_notifications: ^0.0.6+10
awesome_notifications:
     git:
       url: https://github.com/rafaelsetragni/awesome_notifications.git
       ref: 009770e
```

**Comment by husen-hn** [comment]
I don't know why when I set `requireInputText: true`, I can't receive `SilentAction`

```Dart
// receive silent action on Terminated state
NotificationActionButton(
            key: 'SILENT_KEY',
            label: 'Do in Background',
            autoDismissible: true,
            notificationActionType: NotificationActionType.SilentAction)
```

```Dart
// don't receive silent action on Terminated state
NotificationActionButton(
            key: 'SILENT_KEY',
            label: 'Do in Background',
            autoDismissible: true,
            requireInputText: true,
            notificationActionType: NotificationActionType.SilentAction)
```

**Comment by rafaelsetragni** [comment]
To see if its running on terminated i recommend you to check the Logcat console on Android Studio. I did it with the new example app and i could receive the input text inside the parameter `actionInput`

```
2021-09-14 06:05:36.518 8513-8513/me.carda.simple_awesome_example D/NetworkSecurityConfig: No Network Security Config specified, using platform default
2021-09-14 06:05:36.519 8513-8513/me.carda.simple_awesome_example D/NetworkSecurityConfig: No Network Security Config specified, using platform default
2021-09-14 06:05:36.681 8513-8513/me.carda.simple_awesome_example I/DartBackgroundExec: Initializing Flutter global instance.
2021-09-14 06:05:36.704 8513-8537/me.carda.simple_awesome_example I/ResourceExtractor: Found extracted resources res_timestamp-1-1631611086971
2021-09-14 06:05:36.738 8513-8513/me.carda.simple_awesome_example I/DartBackgroundExec: Creating background FlutterEngine instance.
2021-09-14 06:05:36.796 8513-8513/me.carda.simple_awesome_example W/awesome_example: type=1400 audit(0.0:1876): avc: denied { read } for name="max_map_count" dev="proc" ino=46902 scontext=u:r:untrusted_app:s0:c122,c256,c512,c768 tcontext=u:object_r:proc_max_map_count:s0 tclass=file permissive=0 app=me.carda.simple_awesome_example
2021-09-14 06:05:36.827 8513-8513/me.carda.simple_awesome_example D/libEGL: loaded /vendor/lib/egl/libEGL_emulation.so
2021-09-14 06:05:36.832 8513-8513/me.carda.simple_awesome_example D/libEGL: loaded /vendor/lib/egl/libGLESv1_CM_emulation.so
2021-09-14 06:05:36.835 8513-8513/me.carda.simple_awesome_example D/libEGL: loaded /vendor/lib/egl/libGLESv2_emulation.so
2021-09-14 06:05:36.892 8513-8513/me.carda.simple_awesome_example D/HostConnection: HostConnection::get() New Host Connection established 0xefe5fc70, tid 8513
2021-09-14 06:05:36.898 8513-8513/me.carda.simple_awesome_example D/HostConnection: HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_native_sync_v2 ANDROID_EMU_native_sync_v3 ANDROID_EMU_native_sync_v4 ANDROID_EMU_dma_v1 ANDROID_EMU_direct_mem ANDROID_EMU_host_composition_v1 ANDROID_EMU_host_composition_v2 ANDROID_EMU_YUV_Cache ANDROID_EMU_async_unmap_buffer ANDROID_EMU_has_shared_slots_host_memory_allocator ANDROID_EMU_sync_buffer_data GL_OES_EGL_image_external_essl3 GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_host_side_tracing ANDROID_EMU_async_frame_commands ANDROID_EMU_gles_max_version_3_0 
2021-09-14 06:05:37.111 8513-8513/me.carda.simple_awesome_example D/EGL_emulation: eglCreateContext: 0xefe56820: maj 3 min 0 rcv 3
2021-09-14 06:05:37.112 8513-8513/me.carda.simple_awesome_example D/EGL_emulation: eglCreateContext: 0xefe56a50: maj 3 min 0 rcv 3
2021-09-14 06:05:37.135 8513-8546/me.carda.simple_awesome_example D/HostConnection: HostConnection::get() New Host Connection established 0xefe57690, tid 8546
2021-09-14 06:05:37.140 8513-8552/me.carda.simple_awesome_example I/flutter: Observatory listening on http://127.0.0.1:37569/DcLSeMFPW3w=/
2021-09-14 06:05:37.151 8513-8546/me.carda.simple_awesome_example D/HostConnection: HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_native_sync_v2 ANDROID_EMU_native_sync_v3 ANDROID_EMU_native_sync_v4 ANDROID_EMU_dma_v1 ANDROID_EMU_direct_mem ANDROID_EMU_host_composition_v1 ANDROID_EMU_host_composition_v2 ANDROID_EMU_YUV_Cache ANDROID_EMU_async_unmap_buffer ANDROID_EMU_has_shared_slots_host_memory_allocator ANDROID_EMU_sync_buffer_data GL_OES_EGL_image_external_essl3 GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_host_side_tracing ANDROID_EMU_async_frame_commands ANDROID_EMU_gles_max_version_3_0 
2021-09-14 06:05:37.314 8513-8546/me.carda.simple_awesome_example D/EGL_emulation: eglMakeCurrent: 0xefe56a50: ver 3 0 (tinfo 0xf01b1fd0) (first time)
2021-09-14 06:05:37.346 8513-8513/me.carda.simple_awesome_example I/DartBackgroundExec: Executing background FlutterEngine instance.
2021-09-14 06:05:38.027 8513-8513/me.carda.simple_awesome_example D/CompatibilityChangeReporter: Compat change id reported: 147798919; UID 10122; state: ENABLED
2021-09-14 06:05:38.060 8513-8544/me.carda.simple_awesome_example I/flutter: {
2021-09-14 06:05:38.060 8513-8544/me.carda.simple_awesome_example I/flutter:   "actionDate": "2021-09-14 06:05:37",
2021-09-14 06:05:38.060 8513-8544/me.carda.simple_awesome_example I/flutter:   "actionInput": "Test",
2021-09-14 06:05:38.060 8513-8544/me.carda.simple_awesome_example I/flutter:   "actionKey": "REPLY",
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "actionLifeCycle": "AppKilled",
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "autoDismissible": true,
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "backgroundColor": null,
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "bigPicture": null,
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "body": "This is a silent notification to receive notification without bring the app to foreground",
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "channelKey": "background_channel",
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "createdDate": "2021-09-14 06:05:18",
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "createdLifeCycle": "Foreground",
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "createdSource": "Local",
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "defaultColor": null,
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "dismissedDate": null,
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "dismissedLifeCycle": null,
2021-09-14 06:05:38.061 8513-8544/me.carda.simple_awesome_example I/flutter:   "displayedDate": "2021-09-14 06:05:18",
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "displayedLifeCycle": "Foreground",
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "icon": null,
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "id": 1485450034,
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "largeIcon": null,
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "notificationActionType": "SilentAction",
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "payload": null,
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "privacy": "Private",
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "showWhen": true,
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "soundSource": null,
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "summary": null,
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter:   "title": "Silent Notification"
2021-09-14 06:05:38.062 8513-8544/me.carda.simple_awesome_example I/flutter: }
2021-09-14 06:05:38.066 8513-8513/me.carda.simple_awesome_example I/DartBackgroundExec: Shutting down background FlutterEngine instance.
```

Check if you can see your logcat and also try to run the example app to see if the problem persist.

**Comment by husen-hn** [comment]
thank you very much,
the problem I have is that when a message is replayed in terminated mode and the piece of code that is supposed to send that message to the server does not run. but I do not know why the `Fluttertoast` code is executed. (I mean when the app is terminated)

I do not know if this problem is related to this package or not
If not, we can close this issue.

**Comment by AdamBridges** [comment]
> Why youre thinking that im regressing iOS builder? I'm still finishing the iOS implementation.

Hey, thanks for all the hard work, Rafael! 

Do you have any idea when the iOS implementation will be ready?

**Comment by rafaelsetragni** [comment]
Hi! Sorry for taking so long to respond to your issue on GitHub. I was focused on releasing the new 0.7.0 version of `awesome_notifications` and `awesome_notifications_fcm` and i didn't had time enough to do both. But now I can answer all your questions.

So now i'm asking you to recreate this topic using the new issue template. There's a lot of missing informations that i need to understand your problem, and all the instructions are already in the new issue template. Also, remember to check beforehand if your issue was posted by another user.

So I will automatically close all previous issues so far.
Sorry for the inconvenience and i will be waiting for your new issue request.

Thank you so much for your support!
